exonum/src/runtime/execution_context.rs

```diff
@@ -255,36 +279,44 @@ impl ExecutionContextUnstable for ExecutionContext<'_> {
         arguments: &[u8],
         fallthrough_auth: bool,
     ) -> Result<(), ExecutionError> {
-        if self.call_stack_depth + 1 >= Self::MAX_CALL_STACK_DEPTH {
-            let err = CoreError::stack_overflow(Self::MAX_CALL_STACK_DEPTH);
-            return Err(err);
-        }
-
-        let descriptor = self
-            .dispatcher
-            .get_service(called_instance)
-            .ok_or(CoreError::IncorrectInstanceId)?;
-
-        let (runtime_id, runtime) = self
-            .dispatcher
-            .runtime_for_service(descriptor.id)
-            .ok_or(CoreError::IncorrectRuntime)?;
+        // Move logic to closure to catch execution errors in uniform way.
+        let child_call = || {
+            if self.call_stack_depth + 1 >= Self::MAX_CALL_STACK_DEPTH {
+                let err = CoreError::stack_overflow(Self::MAX_CALL_STACK_DEPTH);
+                return Err(err);
```

Comment:
Sorry for not discussing this before, but I wonder whether marking the fork as invalid here really makes sense. Logically, we want to mark the fork as failed if a service returns an error from a call expecting that changes during the call will be rolled back. There is no service call happening here, so (at least theoretically) not marking the fork as failed should be safe.